simulator/risc_v/riscv_decoder.h

```diff
@@ -63,40 +110,115 @@ struct RISCVInstrDecoder
         switch (subset) {
         case 'I': return I_imm;
         case 'B': return get_B_immediate();
-        case 'S': return S_imm4_0 | (S_imm11_5 << 5);
+        case 'S': return S_imm4_0 | (S_imm11_5 << 5U);
         case 'U': return U_imm;
         case 'J': return get_J_immediate();
         case 'C': return csr_imm;
         case ' ': return 0;
-        default: assert(0); return 0;
+        default: return get_compressed_immediate_value( subset);
+        }
+    }
+
+    uint32 get_compressed_immediate_value( char subset) const noexcept
+    {
+        switch (subset) {
+            case C_LWSP: return ( apply_mask( Compr_imm12, 0b1) << 5U)
+                              | ( apply_mask( Compr_imm6_2, 0b11100) << 2U)
+                              | ( apply_mask( Compr_imm6_2, 0b00011) << 6U);
+
+            case C_LDSP: return ( apply_mask( Compr_imm12, 0b1) << 5U)
+                              | ( apply_mask( Compr_imm6_2, 0b11000) << 3U)
+                              | ( apply_mask( Compr_imm6_2, 0b00111) << 6U);
+
+            case C_LQSP: return ( apply_mask( Compr_imm12, 0b1) << 5U)
+                              | ( apply_mask( Compr_imm6_2, 0b10000) << 4U)
+                              | ( apply_mask( Compr_imm6_2, 0b01111) << 6U);
+
+            case C_SWSP: return ( apply_mask( Compr_imm12_7, 0b111100) << 2U)
```

Comment:
For instance, I do not understand the logic here from the code. The variable claims that it contains bits [12:7] of the immediate, in fact it defines bits [5:2] and [7:6].